fix(oci): oci downloader unavailable missing WithBundleParserOpts method

slonka opened the pull request:
<!--

Thanks for submitting a PR to OPA!

Before pressing 'Create pull request' please read the checklist below.

* All code changes should be accompanied with tests. If you are not
modifying any tests, just provide a short explanation of why updates
to tests are not necessary. In addition to helping catch bugs, tests
are extremely helpful in providing _context_ that explains how your
changes can be used.

* All changes to public APIs **must** be accompanied with
docs. Examples of public APIs include built-in functions,
config fields, and of course, exported Go types/functions/constants/etc.

* Commit messages should explain _why_ you made the changes, not what
you changed. Use active voice. Keep the subject line under 50
characters or so.

* All commits must be signed off by the author. If you are not
familiar with signing off, see our contributor guide below.

For more information on contributing to OPA see:

* [Contributing Guide](https://www.openpolicyagent.org/docs/latest/contributing/)
  for high-level contributing guidelines and development setup.

-->

### Why the changes in this PR are needed?

In https://github.com/open-policy-agent/opa/pull/6159 I introduced a disabled OCI downloader to get rid of dependency on containerd to minimise the CVE surface. https://github.com/open-policy-agent/opa/pull/6521/files#diff-f808512360f39e5c4f67342ca24d84117be5403637b032d67e82df3a425cdde9R139 broke our builds adding a method only to one of the implementations. 

<!--
Include a short description of WHY the changes were made.
-->

### What are the changes in this PR?

In this PR I added an interface (should show up in IDE) and a CI check (makes sure building `make GO_TAGS="-tags=opa_no_oci" build` works) to see if everything builds with `GO_TAGS="-tags=opa_no_oci"`.

<!--
Include a short description of WHAT changes were made.
-->

### Notes to assist PR review:

There supposed to be a long term solution for (getting rid of containerd) this but it seems that it was not implemented.

<!--
Here you can add information you think will help the reviewer(s).
-->

### Further comments:

Hope you don't mind me doing a PR - if you want to move this to an issue and discuss - feel free to do so, or tell me to do it :) 

<!--
Here you can include links to additional resources related to the changes, discuss your solution, other approaches you considered etc.
-->

ashutosh-narkar reviewed (commented):
Thanks for working on this @slonka! Instead of introducing an interface if we just implemented `WithBundleParserOpts` in the no OCI downloader this should work, correct? If yes, we can just do that and look into adding a new interface if needed in the future. WDYT?

Also can you please look into the test failures. Thanks.

ashutosh-narkar commented on .github/workflows/pull-request.yaml:
```diff
@@ -30,7 +30,7 @@ jobs:
 
   go-build:
-    name: Go Build (${{ matrix.os }}${{ matrix.arch && format(' {0}', matrix.arch) || '' }})
+    name: Go Build (${{ matrix.os }}${{ matrix.arch && format(' {0}', matrix.arch) || '' }}${{ matrix.go_tags }})
```
What happens in the case `go_tags` is not set?

srenatus replied:
It'll end up as nothing, I think. The CI run of this PR shows it. The only problem is that we seem to be unsetting some args what we need, too: https://github.com/open-policy-agent/opa/actions/runs/7802363034/job/21279888203

slonka replied:
Yes, it will just be empty. Adjusted the call so that it won't interfere with other tags.

ashutosh-narkar replied:
@srenatus does this address your concern?

srenatus replied:
Tests pass, so I'm happy 😃 Thanks!

srenatus commented on download/oci_download_unavailable.go:
```diff
@@ -51,3 +55,7 @@ func (d *OCIDownloader) Start(context.Context) {
 	panic("built without OCI support")
 }
+
+func (d *OCIDownloader) WithBundleParserOpts(ast.ParserOptions) *OCIDownloader {
```
[nit]
```suggestion
func (*OCIDownloader) WithBundleParserOpts(ast.ParserOptions) *OCIDownloader {
```

slonka commented:
> Instead of introducing an interface if we just implemented WithBundleParserOpts in the no OCI downloader this should work, correct? If yes, we can just do that and look into adding a new interface if needed in the future. WDYT?

Yeah, it's not strictly needed but it might show up in some IDEs that a method is not implemented on this struct. A definite guard is what I've added in CI. I'll remove and we can add it in the future.

ashutosh-narkar reviewed (approved):
LGTM. Can you please squash your commits and we can get this in.

slonka commented:
@ashutosh-narkar squashed, can you re-approve so I can merge?

slonka commented:
@ashutosh-narkar do you think we could get this released soon-ish?

ashutosh-narkar commented:
@slonka we have a monthly release cadence. We'll have an OPA release end of this month.
